### PR TITLE
Ayy Laser Guns and UFO Vault Update

### PR DESCRIPTION
**Pull request by Optimism333**
### AYY LASER BEEMS ###
I feel like I aged a few weeks in a few days, but hopefully this pr will have been worth the wait. This pr adds two ayy-themed laser guns to the code. They have unique sprites (thanks to Blithering for help and others for feedback), unique firing modes, unique sound effects, and a unique effect on the death of a target (I'll explain that some more later). 

## Meet the Disintegrator: ##
Disintegrator
DisintegratorInhand

The first and smaller of the guns. Can fire ten shots before it needs a recharge and has scorch and microwave firing modes. It can be dual wielded.

**Scorch Mode:** Very similar to a standard laser beam. It does 30 damage per hit, and you point at things with it and click on them until they fall over from burn damage.

**Microwave Mode:** It's a projectile instead of a laser, and it only deals half the damage of the scorch mode on hit. However, it will raise the target's body temperature significantly upon a successful hit, such that it will deal about 65 total damage over time.

## Meet the HEAVY Disintegrator: ##
heavydisintegrator
HeavyDisintegratorInhand

The bigger brother of the smaller gun. Can fire twenty shots before it needs a recharge and has immolate and scramble firing modes. It also has a firing delay like the laser cannon, though not as extreme. It can't be dual wielded.

**Immolate Mode:** A more powerful laser beam. It does 45 damage per hit, and you point at things with it and click on them until they fall over from burn damage.

**Scramble Mode:** It's a projectile instead of a laser, and it functions like a more nuanced taser. Upon hitting a human target it will instantly make them vomit (effectively a ~6 second stun) and drop what they're holding. It will also deal a very tiny amount of brain damage.

## Unique Effect on Death: ##
Upon annihilation of fleshy hostile alien life forms dial up disintegrator to "well done" and blast target that has ceased life functions to structurally fragment its connective tissue and skeletal structure, leaving behind nothing but ionized fragments.

Or if you'd rather see some visual examples:

https://user-images.githubusercontent.com/81007356/146670315-1b261e82-48a3-488f-be66-c9490a38cb7c.mp4

https://user-images.githubusercontent.com/81007356/146670329-a2d70fe9-1389-4ef7-aa11-22bcb9150ce5.mp4

https://user-images.githubusercontent.com/81007356/146670332-640a4059-58dc-4103-88eb-155bd9a1625b.mp4

The effect "disintegrates" the body of the target upon their death, though the head remains intact for the possibility of revival. The current sanity checks are that it can only occur if the species has blood and if they are dead. It also features unique animations for humans, greys, vox, and insectoids.

## Where can I find one? ##
Both of the guns can be randomly acquired when a wizard summons guns for the crew. However, the regular disintegrator has also been added as a loot item to the grayy ufo vault. I also revamped that vault a little in general to shape it more like a proper round ufo.

brokeufoupdate

You might notice that the grayys in the ufo seem a little different. They ARE a little different. They're more dangerous than before!

GrayyHostiles

There's three new grayy hostile enemy types instead of just the single one from before that could only punch you.

I'll happily elaborate on what the three enemy types are if anyone asks, but maybe you can get an idea from looking at them. And why spoil the fun?

## Closing Thoughts ##
Hopefully the git crowd likes the pr content. As always I'm open to feedback and well-reasoned requests. This one took quite a bit of work. I'm going back to sleep now.

Addendum: Thanks to the coders, spriters, and just general opinion givers who helped out with this pr. I like to think it all made it a bit better.

[balance]  [vault]  [content] [sprites] [sound]

:cl:
 * rscadd: Adds two new grayy themed laser guns
 * rscadd: Revamped layout of broke ufo vault and added new loot
 * rscadd: Grayy hostile humanoids are more dangerous

### CODE REVIEW

**Comment by SonixApache** [comment]
Love it.

**Comment by ManopolisJones** [comment]
But what about the never ending flare?

**Comment by Optimism333** [comment]
> But what about the never ending flare?

Still there, not removed by any means.

**Comment by Eneocho** [comment]
I'd remove the brain damage from the Heavy Disintegrator's scramble mode. Going retarded is way too crippling and can't be easily fixed unlike brute/burn damage.

**Comment by Eneocho** [comment]
Does the "turn players into skeletons" part mean they go husk? Or do they race change to skeleton (such as when cloning they come out as skelemen)?
I can't see the video that well.

**Comment by Optimism333** [comment]
> I'd remove the brain damage from the Heavy Disintegrator's scramble mode. Going retarded is way too crippling and can't be easily fixed unlike brute/burn damage.

It's incredibly small, only 2% damage per shot. I'm not entirely sure when the bad stuff from brain damage kicks in, but I'm pretty sure you'd have to be shot several times with scrambler rays before bad brain damage effects start manifesting. And it can be easily undone with a hit of alkysine or alkycosine.

I tested it on a dummy and I had to shoot him several times before a health analyzer gave the concussion reading.

**Comment by Optimism333** [comment]
> Does the "turn players into skeletons" part mean they go husk? Or do they race change to skeleton (such as when cloning they come out as skelemen)? I can't see the video that well.

It plays an ashing animation and turns them into skeletons, yeah.

**Comment by Eneocho** [comment]
> > I'd remove the brain damage from the Heavy Disintegrator's scramble mode. Going retarded is way too crippling and can't be easily fixed unlike brute/burn damage.
> 
> It's incredibly small, only 2% damage per shot. I'm not entirely sure when the bad stuff from brain damage kicks in, but I'm pretty sure you'd have to be shot several times with scrambler rays before bad brain damage effects start manifesting. And it can be easily undone with a hit of alkysine or alkycosine.

That's the thing, since it otherwise only stuns it can be mass shot until you empty the thing and get the guy retarded. And while alkysine and alkycosine are regularly made, they're not that easy to acquire outside of medical (you can eat syndie cakes or heated donk pockets for that 4 types tricord heal or use the survival box bandage for some quick heal). Also we already have a gun that makes you retarded.

**Comment by Optimism333** [comment]
> > > I'd remove the brain damage from the Heavy Disintegrator's scramble mode. Going retarded is way too crippling and can't be easily fixed unlike brute/burn damage.
> > 
> > 
> > It's incredibly small, only 2% damage per shot. I'm not entirely sure when the bad stuff from brain damage kicks in, but I'm pretty sure you'd have to be shot several times with scrambler rays before bad brain damage effects start manifesting. And it can be easily undone with a hit of alkysine or alkycosine.
> 
> That's the thing, since it otherwise only stuns it can be mass shot until you empty the thing and get the guy retarded. And while alkysine and alkycosine are regularly made, they're not that easy to acquire outside of medical (you can eat syndie cakes or heated donk pockets for that 4 types tricord heal or use the survival box bandage for some quick heal). Also we already have a gun that makes you retarded.

I really don't think it's a problem. The gun is not an effective way to make someone retarded, the effect is simply flavor I added because I imagined that it was "scrambling" the electrical signals in the target's brain and also because while the mode is less than lethal, I didn't want it to be completely harmless.

If a player in theory does use the gun to shoot someone twenty times with that mode then yes, they might be having a bad time. But I'm not really opposed to that when the cure is a common medication from medical. It's no harder to fix than just giving someone a pill and waiting a while, or putting them in the TOOBS if there's a halfway decent mix.

Edit: I just tested it and even after shooting myself twenty times it wasn't enough to become retarded. Only some of the bad effects of brain damage manifested, like dropping things occasionally and eye blurriness. So someone would have to shoot an entire charge into you and then recharge again to keep shooting you if they actually wanted to make you retarded.

**Comment by Eneocho** [comment]
Fair enough. Thanks for testing the issue as well.

**Comment by Anton-Kr** [comment]
https://youtu.be/VBT68a56s70

**Comment by Kurfursten** [comment]
What's going on with that middle?

**Comment by Optimism333** [comment]
> What's going on with that middle?

I'm not sure I understand the question. Can you specify?

Edit: If you mean the middle of the ufo, there's a tile of plasmaglass flooring there. I think it looks nice, since it lets you peek out of the ufo and into space. If you didn't mean that, I don't know.

**Comment by Kurfursten** [comment]
Yep that's what I meant

**Comment by Eneocho** [comment]
Add a fluff only console named abductor ray console near the glass tile, for that classical tractor beam experience.

**Comment by Optimism333** [comment]
> Add a fluff only console named abductor ray console near the glass tile, for that classical tractor beam experience.

If I ever get around to adding ayy machinery/consoles I'd want it to be more than fluff. I've actually thought about adding a "mothership communications" console to the end room of the grayy vault I want to make in the future. Similar to the communications console, but can send messages to the grayy mothership (admin alert like a Centcomm transmission). It'd be interesting to see what the mins and players do with it.

But anyways, I'd want to actually put effort into the console if I added one here, and there's already a lot in the pr. Another time, unless someone else wants to do it in the meantime.

**Comment by Optimism333** [comment]
On further consideration and testing, I think the skeleton death effect still needs some tweaks. Some issues crept up, such as there not being a sanity check for if the dead mob's head is missing (so they'd regrow a head when they turned into a skeleton), and insectoids getting skeletonized doesn't make much sense either (they should get ashed entirely and maybe drop their head since their skeleton is on the outside).

I'm going to convert this into a draft and work on it some more.

**Comment by Optimism333** [comment]
Insectoids will now have their entire body ash instead of getting skellified, but their heads will pop off (so that they may be revived)

https://user-images.githubusercontent.com/81007356/146277382-4684ad7c-0bd8-41d1-9a06-f088e54f0d31.mp4

I know it looks goofy with the human ashing animation, but there's no species-specific ashing animations in-game at the moment.

**Comment by Eneocho** [comment]
Instead of ashing, make the insectoid go husk instead, with those retarded level burns. Ashing and thus demanding a head transplant feels weird when you consider all other victims would be clonable from the get-go.
Does the spookyfication turn vox into skelevox or just regular vox?

**Comment by Optimism333** [comment]
> Instead of ashing, make the insectoid go husk instead, with those retarded level burns. Ashing and thus demanding a head transplant feels weird when you consider all other victims would be clonable from the get-go. Does the spookyfication turn vox into skelevox or just regular vox?

I'm considering applying the insectoid effect to all the fleshy species, tbh. 

While I love the look of the skellington disintegration/transformation, it's currently very gunked because I don't know how to make it check for if the victim had a head and not spawn a head if the victim didn't have one. At minimum I'd like to apply enough sanity such that it doesn't regrow the head of a headless corpse.

It's weird enough that it would regrow missing limbs, but I'd at least like to make it not regrow heads.

**Comment by Optimism333** [comment]
> Does the spookyfication turn vox into skelevox or just regular vox?

Vox currently turn into skelevox

**Comment by Eneocho** [comment]
I can't help you on the sanitychecks, but you could look at the cloner code for ideas. It checks for a head with a brain. Or look at the surgery code since it has checks for limbs (you cant behead a beheaded corpse, for example).

**Comment by Optimism333** [comment]
> I can't help you on the sanitychecks, but you could look at the cloner code for ideas. It checks for a head with a brain. Or look at the surgery code since it has checks for limbs (you cant behead a beheaded corpse, for example).

Thanks, I'll look later. 

In an ideal outcome, I'll be able to make it so that the species change inherits any missing limbs, but if not I'll just see how much I can do.

**Comment by Blithering** [comment]
Cheeky way to change it to be simpler is to make the skeleton effect trigger if the victim is ALIVE and has >X damage, this would bypass the need for a lot of sanity since people can't live headless for long, and could also mean cool standing skeletonisation.

**Comment by Optimism333** [comment]
Unfortunately I have to scrap the skeleton species transformation for now, since I just don't see it getting done, at least not within a reasonable amount of time. The way species transformations are coded is beyond my comprehension and I'd be more likely to just break things if I tried to alter it. I tried a few different methods and nothing worked, so for the time being I'm going to move forward with a solution I know I can actually implement.

The disintegrator's on-death effect will appear as in the videos below:

https://user-images.githubusercontent.com/81007356/146670040-4b5a1f33-2e1e-4b71-95dd-b0f93981938c.mp4

https://user-images.githubusercontent.com/81007356/146670050-2eae8bcd-a65f-4c9f-839e-4eca17a31202.mp4

https://user-images.githubusercontent.com/81007356/146670056-be338ef6-946a-47a2-b1c7-496efa449ca1.mp4

https://user-images.githubusercontent.com/81007356/146670060-53e54405-d62e-4010-89a0-e6eb44cce384.mp4

The gun will destroy the target's body on death while playing an animation, but it will drop the head so that the player may still be revived. A bit of consolation is that I spent time spriting standing and lying ashing animations for vox, insectoids, greys, and humans. That way the effect looks visually decent, even if the mechanic won't be as impressive as I wanted it to be. I'll probably add more ashing animations for rarer species like mushmen, catbeasts, unathi, and skrell when I have some time.

**Comment by Eneocho** [comment]
You could delete the original mob, spawn in a new skeleton mob with the same name, remove limbs if appropriate then move the mind to the new skeleton mob.
Not sure how you'd do that though.

**Comment by Optimism333** [comment]
> You could delete the original mob, spawn in a new skeleton mob with the same name, remove limbs if appropriate then move the mind to the new skeleton mob. Not sure how you'd do that though.

Yeah, I considered that. I even tried using an old pr that zth started as a guideline (an old one that was supposed to turn people who took too much burn damage from plasmafires into plasmamen), and that was the closest I got to actually pulling it off. But unfortunately the result was a skellington that didn't run fast and could bleed, even though it respected lost limbs. (I also couldn't get it to turn vox into skelevox when I used that option)

At this point I just think it's better to do something I actually know how to do, rather than hold back this pr until I figure out a way to make the species transformation work.

File: code/modules/mob/living/simple_animal/hostile/human/grey.dm [hunk 1]
```diff
@@ -20,18 +33,75 @@
...
+	melee_damage_upper = 18
+
+	items_to_drop = list(/obj/item/tool/scalpel)
+
+	speak = list("I need a new specimen to dissect.","A dissection is all I need... just one more dissection.","Praise the mothership.")
+	speak_chance = 5
+
+/mob/living/simple_animal/hostile/humanoid/grey/space/melee/scalpel/Aggro()
+	..()
+	say(pick("I need your organs for testing!","You'll make a fine specimen for an operation!"))
```
Comment (Kurfursten) [inline code/modules/mob/living/simple_animal/hostile/human/grey.dm]: if you want to make these guys speak quacks, you can just add this after your pick:

`, get_language(LANGUAGE_GREY))`

File: code/modules/projectiles/guns/energy/disintegrator.dm [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	modifystate = "heavydisintegratorimmolate"
+	fire_delay = 1.2 SECONDS // Here to slightly counterbalance the more damaging ray, but a lot less noticeable than the laser cannon
+
+	var/mode = 0 //0 = immolate, 1 = scramble
+
+/obj/item/weapon/gun/energy/heavydisintegrator/attack_self(mob/living/user as mob)
+	switch(mode)
+		if(0)
+			mode = 1
+			charge_cost = 50
```
Comment (Kurfursten) [inline code/modules/projectiles/guns/energy/disintegrator.dm]: why are you setting charge cost if it doesn't change?

File: code/modules/mob/living/simple_animal/hostile/human/grey.dm (hunk 1 above)
Comment (Optimism333) [reply code/modules/mob/living/simple_animal/hostile/human/grey.dm]: Tried it, and it seems to throw back an error saying that get_language is an undefined proc.

Capture
 
I really would like to make these guys speak grey if I can, that would be an excellent detail.

File: code/modules/projectiles/guns/energy/disintegrator.dm (hunk 2 above)
Comment (Optimism333) [reply code/modules/projectiles/guns/energy/disintegrator.dm]: Inexperience led to difficulty separating necessities from redundancies. I was mostly looking at how the nuclear.dm gun page handled it. Either way, all of the unnecessary extra charge cost lines have been removed.

File: code/modules/mob/living/simple_animal/hostile/human/grey.dm (hunk 1 above)
Comment (Optimism333) [reply code/modules/mob/living/simple_animal/hostile/human/grey.dm]: I'm guessing that means I have to add a line to the page somewhere that defines said proc. Uncertain how to go about that, though.

Comment (Kurfursten) [reply code/modules/mob/living/simple_animal/hostile/human/grey.dm]: no need, I forgot that proc isn't universal

Instead do this:
`, all_languages[LANGUAGE_GREY]))`

all_languages is a global list so it should be good
I also suggest you add this to the `/humanoid/grey/New()` --
```
languages += all_languages[LANGUAGE_GREY]
```

This should ensure that if someone is bussed into one of these mobs, they will understand grey quacks

Comment (Optimism333) [reply code/modules/mob/living/simple_animal/hostile/human/grey.dm]: Okay, I made an attempt to apply the changes in a new commit. Let me know if they look alright or still need to be refined. Tested them and it seemed to work.

**Pull request merged**